### PR TITLE
Replace all uses of deprecated symbol_exprt constructors [blocks: #3768]

### PR DESCRIPTION
**Pull request by tautschnig**
This helps type safety has it avoids constructing symbol_exprts that never get
a proper type (or identifier).

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 115efb34e0442462fd5cef19dc2cd0be2f4596e1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97091511

File: jbmc/src/java_bytecode/simple_method_stubbing.cpp [hunk 1]
```diff
@@ -170,7 +170,8 @@ void java_simple_method_stubst::create_method_stub(symbolt &symbol)
       symbol_table);
     const symbol_exprt &init_symbol_expression = init_symbol.symbol_expr();
     code_assignt get_argument(
-      init_symbol_expression, symbol_exprt(this_argument.get_identifier()));
+      init_symbol_expression,
+      symbol_exprt(this_argument.get_identifier(), this_argument.type()));
     get_argument.add_source_location() = synthesized_source_location;
```
Comment (kroening) [inline jbmc/src/java_bytecode/simple_method_stubbing.cpp]: This one does appear to be a bugfix!

File: jbmc/unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp [hunk 2]
```diff
@@ -215,7 +215,7 @@ SCENARIO("instantiate_not_contains",
         simplify(sc.premise, empty_ns);
         simplify(sc.s0, empty_ns);
         simplify(sc.s1, empty_ns);
-        witnesses[sc] = generator.fresh_symbol("w", t.witness_type());
+        witnesses.emplace(sc, generator.fresh_symbol("w", t.witness_type()));
         nc_axioms.push_back(sc);
```
Comment (kroening) [inline jbmc/unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp]: Not sure that the .emplace adds a lot here (exprt is very cheap to construct).

File: src/analyses/locals.cpp [hunk 3]
```diff
@@ -21,12 +21,13 @@ void localst::build(const goto_functiont &goto_function)
...
+      locals_map.emplace(code_decl.get_identifier(), code_decl.symbol());
     }
 
   for(const auto &param : goto_function.type.parameters())
-    locals_map[param.get_identifier()]=
-      symbol_exprt(param.get_identifier(), param.type());
+    locals_map.emplace(
+      param.get_identifier(),
+      symbol_exprt(param.get_identifier(), param.type()));
 }
```
Comment (kroening) [inline src/analyses/locals.cpp:31]: Same here

File: src/goto-instrument/concurrency.cpp [hunk 4]
```diff
@@ -58,14 +58,14 @@ class concurrency_instrumentationt
   {
   public:
     typet type;
-    symbol_exprt array_symbol, w_index_symbol;
+    exprt array_symbol, w_index_symbol;
   };
```
Comment (kroening) [inline src/goto-instrument/concurrency.cpp]: Can these ever be something that's not a symbol_exprt?
If initialisation is delayed, then perhaps optionalt<symbol_exprt> is the right way to say that.

File: src/goto-symex/symex_function_call.cpp [hunk 5]
```diff
@@ -409,7 +409,7 @@ void goto_symext::locality(
 
     if(c_it != state.level1.current_names.end())
     {
-      frame.old_level1[l0_name]=c_it->second;
+      frame.old_level1.emplace(l0_name, c_it->second);
       c_it->second = std::make_pair(ssa, frame_nr);
```
Comment (kroening) [inline src/goto-symex/symex_function_call.cpp:412]: Same here

File: src/util/ssa_expr.h [hunk 6]
```diff
@@ -16,15 +16,9 @@ Author: Daniel Kroening, kroening@kroening.com
...
-  {
-    set(ID_C_SSA_symbol, true);
-  }
-
   /// Constructor
   /// \param expr: Expression to be converted to SSA symbol
-  explicit ssa_exprt(const exprt &expr):
-    symbol_exprt(expr.type())
+  explicit ssa_exprt(const exprt &expr) : symbol_exprt(ID_nil, expr.type())
   {
```
Comment (kroening) [inline src/util/ssa_expr.h:22]: This is the first time you've used the nil ID -- before it has been the empty one.
If you are concerned about any time taken to convert "" into an irep_idt, then use irep_idt().

**Review by kroening — APPROVED** [review]

File: jbmc/unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp (hunk 2 above)
Comment (tautschnig) [reply jbmc/unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp]: The `.emplace` is necessary, because a `symbol_exprt` cannot be default constructed.

File: src/util/ssa_expr.h (hunk 6 above)
Comment (tautschnig) [reply src/util/ssa_expr.h:22]: Now using `irep_idt()` here.

File: jbmc/src/java_bytecode/simple_method_stubbing.cpp (hunk 1 above)
Comment (tautschnig) [reply jbmc/src/java_bytecode/simple_method_stubbing.cpp]: Yes, but it should actually get merged via #3747, I'll then rebase this PR on top of it.

File: src/goto-instrument/concurrency.cpp (hunk 4 above)
Comment (tautschnig) [reply src/goto-instrument/concurrency.cpp]: Well, the code is not particularly complete. These are actually never set at the moment.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2a5c14aa579799c03bb2b55362c6c1717b5994e5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97291629

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b71d8f91adbb941982c5970135d0ea5dec581bee).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97316228

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 7]
```diff
@@ -285,7 +285,8 @@ exprt make_function_application(
   declare_function(fun_name, type, symbol_table);
 
   // Function application
-  return function_application_exprt(symbol_exprt(fun_name), arguments, type);
+  return function_application_exprt(
+    symbol_exprt(fun_name, typet()), arguments, type);
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_utils.cpp]: This function doesn't seem correct. I think it should be changed to take a `code_typet` as argument representing the type of the function instead of `type` (the return type can be deduced from the `code_typet`

File: jbmc/src/java_bytecode/java_bytecode_convert_method_class.h [hunk 8]
```diff
@@ -119,7 +119,8 @@ class java_bytecode_convert_methodt:public messaget
     bool is_parameter;
     std::vector<holet> holes;
 
-    variablet() : symbol_expr(), start_pc(0), length(0), is_parameter(false)
+    variablet()
+      : symbol_expr("", typet()), start_pc(0), length(0), is_parameter(false)
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_convert_method_class.h]: Would it make sense for the type to be an argument of the constructor?

File: src/goto-symex/slice_by_trace.cpp [hunk 9]
```diff
@@ -31,10 +31,6 @@ void symex_slice_by_tracet::slice_by_trace(
 {
   std::cout << "Slicing by trace...\n";
 
-  merge_identifier="goto_symex::\\merge";
-  merge_symbol=symbol_exprt(typet(ID_bool));
-  merge_symbol.set_identifier(merge_identifier);
-
```
Comment (romainbrenguier) [inline src/goto-symex/slice_by_trace.cpp]: This should have been in a separate commit, so that you can explain this specific change

File: src/goto-symex/slice_by_trace.h [hunk 10]
```diff
@@ -17,9 +17,11 @@ Author: Alex Groce (agroce@gmail.com)
...
 {
 public:
-  explicit symex_slice_by_tracet(const namespacet &_ns):
-    ns(_ns),
-    alphabet_parity(false)
+  explicit symex_slice_by_tracet(const namespacet &_ns)
+    : ns(_ns),
+      alphabet_parity(false),
+      merge_symbol("goto_symex::\\merge", typet(ID_bool))
+
```
Comment (romainbrenguier) [inline src/goto-symex/slice_by_trace.h]: extra blank line

File: src/solvers/smt2/smt2_parser.h [hunk 11]
```diff
@@ -46,6 +46,11 @@ class smt2_parsert
 
   struct named_termt
   {
+    named_termt(const exprt &_term, const symbol_exprt &_name)
+      : term(_term), name(_name)
+    {
+    }
+
```
Comment (romainbrenguier) [inline src/solvers/smt2/smt2_parser.h:56]: should have been in a separate commit to explain why this change is needed

**Review by romainbrenguier — COMMENTED** [review]
Would have been better if split into several commits. `typet()` is used in several places so I'm not sure how this changes are useful.

**Comment by smowton** [comment]
I get that you want to avoid people casually saying `symbol_exprt("hello")` and stumble into a dangerous part-constructed expression without noticing -- but as Romain notes, this is just being replaced with explicit `typet()`. That being so, how about moving the evil constructor behind a static function? For example:

```
private:
  // Partial constructor, to be avoided where possible. Use via `::typeless`
  symbol_exprt(const irep_idt &id) { ... }

public:
  // Prefer the public constructor whenever possible.
  static symbol_exprt typeless(const irep_idt &id) { return symbol_exprt(id); }
```

File: jbmc/src/java_bytecode/java_utils.cpp (hunk 7 above)
Comment (tautschnig) [reply jbmc/src/java_bytecode/java_utils.cpp]: So I would have thought, but this does not seem to be the case for a `function_application_exprt`, or else what the Java front-end is doing is completely broken. @kroening maybe?

**Comment by tautschnig** [comment]
I'll try to implement @smowton's suggestion as soon as possible.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ab2ae0ebfede21b2cbede41f101fb79e269c855c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98019728

**Comment by tautschnig** [comment]
I have started to refactor the commit into smaller ones and have created first PRs that just contain those.

**Comment by tautschnig** [comment]
Separate PRs created where the individual bits should now cover all the feedback received.

**Comment by tautschnig** [comment]
This PR now only has a single commit left, (re-)reviews would be appreciated.

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c539028f83d56857b54e5838958b189d952d5cc3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98341703

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 45c74ceb269b65e34441e6e3785ad993a31e2c3d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98712690

File: src/solvers/smt2/smt2_parser.h (hunk 11 above)
Comment (martin-cs) [reply src/solvers/smt2/smt2_parser.h:56]: +1 this is not clear.

**Review by martin-cs — APPROVED** [review]
The bits that touch my areas of responsibility seem fine.

File: src/solvers/smt2/smt2_parser.h (hunk 11 above)
Comment (tautschnig) [reply src/solvers/smt2/smt2_parser.h:56]: I have added a comment (default-constructing a `symbol_exprt` is deprecated, thus the `name` member needs to be initialised upon construction). Goes together with the use of `emplace` in smt2_parser.cpp.

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: d16fbebca858fd37275d17025d5a88567c421c1b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98896786
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 267094c7eb4bd9cca8b3bcae22e60a9b67d86a7f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99079556
